tests/test_command.py

```diff
@@ -555,14 +555,144 @@ def test_stub(self):
+        print_command()
+        Connection._api_call = Mock(return_value={'id': 1234})
+        qds.main()
+        Connection._api_call.assert_called_with('POST', 'commands',
+                {'export_dir': None,
+                 'name': 'commandname',
+                 'db_update_keys': None,
+                 'partition_spec': None,
+                 'fields_terminated_by': None,
+                 'hive_table': 'myhivetable',
+                 'db_table': 'mydbtable',
+                 'mode': '1',
+                 'tags': None,
+                 'command_type': 'DbExportCommand',
+                 'dbtap_id': '1',
+                 'can_notify': False,
+                 'db_update_mode': None})
+
+    def test_submit_with_update_mode_and_keys(self):
+        sys.argv = ['qds.py', 'dbexportcmd', 'submit', '--mode', '1', '--dbtap_id', '1',
+         '--db_table', 'mydbtable', '--hive_table', 'myhivetable',
+          '--db_update_mode', 'updateonly', '--db_update_keys', 'key1']
+        print_command()
+        Connection._api_call = Mock(return_value={'id': 1234})
+        qds.main()
+        Connection._api_call.assert_called_with('POST', 'commands',
+                {'export_dir': None,
+                 'name': None,
+                 'db_update_keys': 'key1',
+                 'partition_spec': None,
+                 'fields_terminated_by': None,
+                 'hive_table': 'myhivetable',
+                 'db_table': 'mydbtable',
+                 'mode': '1',
+                 'tags': None,
+                 'command_type': 'DbExportCommand',
+                 'dbtap_id': '1',
+                 'can_notify': False,
+                 'db_update_mode': 'updateonly'})
+
+    def test_submit_with_mode_2(self):
+        sys.argv = ['qds.py', 'dbexportcmd', 'submit', '--mode', '2', '--dbtap_id', '1',
+         '--db_table', 'mydbtable', '--hive_table', 'myhivetable',
+           '--export_dir', 's3:///export-path/']
+        print_command()
+        Connection._api_call = Mock(return_value={'id': 1234})
+        qds.main()
+        Connection._api_call.assert_called_with('POST', 'commands',
+                {'export_dir': 's3:///export-path/',
+                 'name': None,
+                 'db_update_keys': None,
+                 'partition_spec': None,
+                 'fields_terminated_by': None,
+                 'hive_table': 'myhivetable',
+                 'db_table': 'mydbtable',
+                 'mode': '2',
+                 'tags': None,
+                 'command_type': 'DbExportCommand',
+                 'dbtap_id': '1',
+                 'can_notify': False,
+                 'db_update_mode': None})
 
 
 class TestDbImportCommand(QdsCliTestCase):
 
-    def test_stub(self):
-        pass
+    # Not much point adding more test cases as the semantic check in main code is still remaining.
+    # The test cases might give false positivies
```

Comment:
So here is the parse method for DbExportCommand:

def parse(cls, args):
        """       Parse command line arguments to construct a
dictionary of command       parameters that can be used to create a
command       Args:           `args`: sequence of arguments
Returns:           Dictionary that can be used in create method
Raises:           ParseError: when the arguments are not correct
"""

        try:
            (options, args) = cls.optparser.parse_args(args)
            if options.mode not in ["1", "2"]:
                raise ParseError("mode must be either '1' or '2'",
                                 cls.optparser.format_help())

            if (options.dbtap_id is None) or (options.db_table is None):
                raise ParseError("dbtap_id and db_table are required",
                                 cls.optparser.format_help())

            if options.mode is "1":
                if options.hive_table is None:
                    raise ParseError("hive_table is required for mode 1",
                                     cls.optparser.format_help())
            elif options.export_dir is None:    # mode 2
                raise ParseError("export_dir is required for mode 2",
                                 cls.optparser.format_help())

            if options.db_update_mode is not None:
                if options.db_update_mode not in ["allowinsert", "updateonly"]:
                    raise ParseError("db_update_mode should either be
left blank for append "
                                     "mode or be 'updateonly' or 'allowinsert'",
                                     cls.optparser.format_help())
                if options.db_update_mode is "updateonly":
                    if options.db_update_keys is None:
                        raise ParseError("db_update_keys is required
when db_update_mode "
                                         "is 'updateonly'",
                                         cls.optparser.format_help())
                elif options.db_update_keys is not None:
                    raise ParseError("db_update_keys is used only when
db_update_mode "
                                     "is 'updateonly'",
                                     cls.optparser.format_help())

        except OptionParsingError as e:
            raise ParseError(e.msg, cls.optparser.format_help())
        except OptionParsingExit as e:
            return None

        v = vars(options)
        v["command_type"] = "DbExportCommand"
        return v


And here is the parse method for DbImportCommand:

def parse(cls, args):
        """        Parse command line arguments to construct a
dictionary of command        parameters that can be used to create a
command        Args:            `args`: sequence of arguments
Returns:            Dictionary that can be used in create method
 Raises:            ParseError: when the arguments are not correct
   """

        try:
            (options, args) = cls.optparser.parse_args(args)
            if options.mode not in ["1", "2"]:
                raise ParseError("mode must be either '1' or '2'",
                                 cls.optparser.format_help())

            if (options.dbtap_id is None) or (options.db_table is None):
                raise ParseError("dbtap_id and db_table are required",
                                 cls.optparser.format_help())

            # TODO: Semantic checks for parameters in mode 1 and 2

        except OptionParsingError as e:
            raise ParseError(e.msg, cls.optparser.format_help())
        except OptionParsingExit as e:
            return None

        v = vars(options)
        v["command_type"] = "DbImportCommand"
        return v


So, suppose i write a test case with mode 2 and give hive table as a
parameter (dbimportcmd run --mode 2 --hive_table abc), it should
ideally return an error (and the test case should be written
accordingly), but instead the sdk will make an API call since the
semantic check is missing. Hence there was no point in writing test
cases for it now. Maybe we can raise a JIRA to complete the parse
method of DbImportCommand first and then write the test cases for it
too.


On Thu, Jun 4, 2015 at 11:26 PM, Rohit Agarwal <notifications@github.com>
wrote:

> In tests/test_command.py
> <https://github.com/qubole/qds-sdk-py/pull/77#discussion_r31746538>:
>
> >
> >
> >  class TestDbImportCommand(QdsCliTestCase):
> >
> > -    def test_stub(self):
> > -        pass
> > +    # Not much point adding more test cases as the semantic check in main code is still remaining.
> > +    # The test cases might give false positivies
>
> @rohitpruthi95 <https://github.com/rohitpruthi95> I don't understand
> this. Can you give some examples?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qubole/qds-sdk-py/pull/77/files#r31746538>.
>